Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/revert.py

```diff
@@ -0,0 +1,152 @@
+# modification, are permitted provided that the following conditions
+# are met:
+# 1.  Redistributions of source code must retain the above copyright
+#     notice, this list of conditions and the following disclaimer.
+# 2.  Redistributions in binary form must reproduce the above copyright
+#     notice, this list of conditions and the following disclaimer in the
+#     documentation and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS BE LIABLE FOR
+# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import os
+import re
+import sys
+from xml.etree.ElementInclude import include
+
+from .command import Command
+from .branch import Branch
+from.pull_request import PullRequest
+
+from webkitbugspy import Tracker
+from webkitcorepy import arguments, run, Terminal
+from webkitscmpy import local, log, remote
+
+
+class Revert(Command):
+    name = 'revert'
+    help = 'Revert provided list of commits and create a pull-request with this revert commit'
+
+    @classmethod
+    def parser(cls, parser, loggers=None):
+        parser.add_argument(
+            'commits',
+            dest='commits', required= True, nargs="+",
+            help='git hash, svn revision or identifer you want to revert',
+            action=arguments.NoAction,
+        )
+        PullRequest.parser(parser)
+
+    @classmethod
+    def revert_commit(cls, args, repository, **kwargs):
+        # Check if there are any outstanding changes:
+        if repository.modified():
+            sys.stderr.write('Please commit your changes or stash them before you revert commit: {}'.format(args.commits))
+            return 1
+        # Make sure we have the commit that user want to revert
+        try:
+            commits = map(lambda commit_arg: repository.find(commit_arg, include_log=False), args.commits)
+        except (local.Scm.Exception, ValueError) as exception:
+            # ValueErrors and Scm exceptions usually contain enough information to be displayed
+            # to the user as an error
+            sys.stderr.write(str(exception) + '\n')
+            return 1
+
+        commit_hashes = list(map(lambda commit: commit.hash, commits))
+        result = run([repository.executable(), 'revert', '--no-commit'] + commit_hashes, cwd=repository.root_path, capture_output=True)
```

Comment:
Looking through this change, looks like we will need to mock `git revert` and `git checkout` to get unit tests working.